File: FBControlCore/Commands/FBXCTraceRecordCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBXCTraceRecordCommands.h"

#import "FBXCTestShimConfiguration.h"
#import "FBXCTraceConfiguration.h"
#import "FBXCTraceOperation.h"
#import "FBiOSTarget.h"

@implementation FBXCTraceRecordCommands

#pragma mark initializers

+ (instancetype)commandsWithTarget:(id<FBiOSTarget>)target
{
  return [[self alloc] initWithTarget:target];
}

- (instancetype)initWithTarget:(id<FBiOSTarget>)target
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _target = target;

  return self;
}

#pragma mark FBXCTraceRecordCommands

- (FBFuture<FBXCTraceRecordOperation *> *)startXctraceRecord:(FBXCTraceRecordConfiguration *)configuration logger:(id<FBControlCoreLogger>)logger
{
  return [[FBXCTestShimConfiguration
    sharedShimConfigurationWithLogger:logger]
    onQueue:self.target.workQueue fmap:^(FBXCTestShimConfiguration *shim) {
      return [FBXCTraceRecordOperation operationWithTarget:self.target configuration:[configuration withShim:shim] logger:logger];
  }];
}

@end

File: FBControlCore/Commands/FBXCTraceRecordCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public final class FBXCTraceRecordCommands: FBiOSTargetCommand, XCTraceRecordCommands {

  public let target: any FBiOSTarget

  public class func commands(with target: any FBiOSTarget) -> Self {
    self.init(target: target)
  }

  required init(target: any FBiOSTarget) {
    self.target = target
  }

  // MARK: - Operations

  public func start(configuration: XCTraceRecordConfiguration, logger: any FBControlCoreLogger) async throws -> XCTraceRecordOperation {
    let shim = try await XCTestShimConfiguration.sharedShimConfiguration()
    return try await XCTraceRecordOperation.operation(with: target, configuration: configuration.withShim(shim), logger: logger)
  }
}
